### PR TITLE
WooCommerce - added label price breakdown (promoting the discount)

### LINKED ISSUE #19935: Store: Promote shipping label postage savings to merchant
_WCS development is now Calypso-first ✨ , moving https://github.com/Automattic/woocommerce-services/issues/819 here._

To do:

- [ ] Update design for Calypso
- [ ] Remove insurance
- [ ] Add tooltip to "your discount" row w/explanation of retail vs. CBP

### PR DESCRIPTION
**Pull request by marcinbot**
Work in progress

Fixes #19935

Adds label price breakdown:
screen shot 2017-12-01 at 16 42 27

To test:
* in the label purchase modal, select a rate, or rates for multiple boxes
* verify that the component appears and that it's behaviour is correct
* verify that all the prices shown throughout the modal are correct (total on the purchase button, prices next to the rates)

### CODE REVIEW

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	};
+
+	const retailSum = sum( map( priceBreakdown, 'retailRate' ) );
+	const total = sum( map( priceBreakdown, 'rate' ) );
+	const discount = total - retailSum;
+
+	return (
+		<Card>
+			{ priceBreakdown.map( ( service, index ) => renderService( service.title, service.retailRate, index ) ) }
+			{ 0 > discount ? renderService( translate( 'Your discount' ), discount, 'discount' ) : null }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js]: Hi! I've found a possible matching string that has already been translated **41** times:
`translate( 'Discount' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/19935-label-price-breakdown)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		);
+	};
+
+	const { prices, fee, discount, total } = priceBreakdown;
+
+	return (
+		<Card>
+			{ prices.map( ( service, index ) => renderRow( service.title, service.retailRate, index ) ) }
+			{ 0 < fee ? renderRow( translate( 'Labels fee' ), fee, 'fee' ) : null }
+			{ 0 < discount ? renderRow( translate( 'Your discount' ), -discount, 'discount' ) : null }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js]: Hi! I've found a possible matching string that has already been translated **41** times:
`translate( 'Discount' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/19935-label-price-breakdown)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		if ( ! priceBreakdown ) {
+			return null;
+		}
+
+		const { prices, discount, total } = priceBreakdown;
+
+		return (
+			<Card>
+				{ prices.map( ( service, index ) => this.renderRow( service.title, service.retailRate, index ) ) }
+				{ 0 < discount ? this.renderRow( translate( 'Your discount' ), -discount, 'discount', false, true ) : null }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js]: Hi! I've found a possible matching string that has already been translated **41** times:
`translate( 'Discount' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/19935-label-price-breakdown)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Comment (a8ci18n) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js]: Hi! I've found a possible matching string that has already been translated **41** times:
`translate( 'Discount' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/19935-label-price-breakdown)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js [hunk 4]
```diff
@@ -74,33 +74,49 @@ export const getForm = ( state, orderId, siteId = getSelectedSiteId( state ) ) =
...
-		if ( ! form ) {
-			return 0;
-		}
-
-		const { values: selectedRates, available: availableRates } = form.rates;
-
-		const ratesCost = map( selectedRates, ( rateId, boxId ) => {
-			const packageRates = get( availableRates, [ boxId, 'rates' ], false );
+/**
+ * Returns a breakdown of the total price for selected labels in form of { prices, discount, fee, total }
```
Comment (jeffstieler) [inline client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js]: `fee` isn't part of the return.

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js [hunk 5]
```diff
@@ -74,33 +74,49 @@ export const getForm = ( state, orderId, siteId = getSelectedSiteId( state ) ) =
...
-			if ( packageRates ) {
-				const foundRate = find( packageRates, [ 'service_id', rateId ] );
+	const { values: selectedRates, available: availableRates } = form.rates;
+	const prices = [];
+	let discount = 0;
+	let total = 0;
+	for ( const packageId in selectedRates ) {
+		const packageRates = get( availableRates, [ packageId, 'rates' ], false );
+		//if retail_rate is not set on any rate, then the selector should return null
+		if ( ! packageRates || ! packageRates.length || ! isFinite( packageRates[ 0 ].retail_rate ) ) {
```
Comment (jeffstieler) [inline client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js]: If you're interested, you should be able to pass the `0` and `retail_rates` as part of your call to `get()` above.

Like `get( availableRates, [ packageId, 'rates', 0, 'retail_rate' ], false )` (untested)

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+							onMouseLeave={ this.hideTooltip }
+							size={ 18 } />
+						<Tooltip
+							className="label-purchase-modal__price-item-tooltip is-dialog-visible"
+							isVisible={ this.state.tooltipVisible }
+							context={ this.state.tooltipContext } >
+						{ translate( 'WooCommerce Services gives you access to USPS ' +
+							'Commercial Pricing, which is discounted over Retail rates.' ) }
+						</Tooltip>
+					</div> }
```
Comment (jeffstieler) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/price-summary.js]: What do you think about moving the discount tooltip markup to it's own render method to make `renderRow` a little easier to grok?

**Pull request merged**